autotest/gcore/vsifile.py

```diff
@@ -1000,6 +1000,133 @@ def my_progress(pct, message, user_data):
+        return 'fail'
+    entry = gdal.GetNextDirEntry(d)
+    if entry:
+        gdaltest.post_reason('fail')
+        return 'fail'
+    gdal.CloseDir(d)
+
+    gdal.FileFromMemBuffer('/vsimem/vsifile_opendir/test', 'foo')
+    gdal.Mkdir('/vsimem/vsifile_opendir/subdir', 493)
+    gdal.Mkdir('/vsimem/vsifile_opendir/subdir/subdir2', 493)
+    gdal.FileFromMemBuffer('/vsimem/vsifile_opendir/subdir/subdir2/test2', 'bar')
+
+    # Unlimited depth
+    d = gdal.OpenDir('/vsimem/vsifile_opendir')
+
+    entry = gdal.GetNextDirEntry(d)
+    if entry.name != 'subdir':
+        gdaltest.post_reason('fail')
+        print(entry.name)
+        return 'fail'
+    if entry.mode != 16384:
+        gdaltest.post_reason('fail')
+        print(entry.mode)
+        return 'fail'
+
+    entry = gdal.GetNextDirEntry(d)
+    if entry.name != 'subdir/subdir2':
+        gdaltest.post_reason('fail')
+        print(entry.name)
+        return 'fail'
+    if entry.mode != 16384:
+        gdaltest.post_reason('fail')
+        print(entry.mode)
+        return 'fail'
+
+    entry = gdal.GetNextDirEntry(d)
+    if entry.name != 'subdir/subdir2/test2':
+        gdaltest.post_reason('fail')
+        print(entry.name)
+        return 'fail'
+    if entry.mode != 32768:
+        gdaltest.post_reason('fail')
+        print(entry.mode)
+        return 'fail'
+
+    entry = gdal.GetNextDirEntry(d)
+    if entry.name != 'test':
+        gdaltest.post_reason('fail')
+        print(entry.name)
+        return 'fail'
+    if entry.mode != 32768:
+        gdaltest.post_reason('fail')
+        print(entry.mode)
+        return 'fail'
+    if not entry.modeKnown:
+        gdaltest.post_reason('fail')
+        return 'fail'
+    if entry.size != 3:
+        gdaltest.post_reason('fail')
+        return 'fail'
+    if not entry.sizeKnown:
+        gdaltest.post_reason('fail')
+        return 'fail'
+    if entry.mtime == 0:
+        gdaltest.post_reason('fail')
+        return 'fail'
+    if not entry.mtimeKnown:
+        gdaltest.post_reason('fail')
+        return 'fail'
+    if entry.extra:
+        gdaltest.post_reason('fail')
+        return 'fail'
+
+    entry = gdal.GetNextDirEntry(d)
+    if entry:
+        gdaltest.post_reason('fail')
+        return 'fail'
+    gdal.CloseDir(d)
+
+    # Only top level
+    d = gdal.OpenDir('/vsimem/vsifile_opendir', 0)
+    entry = gdal.GetNextDirEntry(d)
+    if entry.name != 'subdir':
+        gdaltest.post_reason('fail')
+        print(entry.name)
+        return 'fail'
+    entry = gdal.GetNextDirEntry(d)
+    if entry.name != 'test':
+        gdaltest.post_reason('fail')
+        print(entry.name)
+        return 'fail'
+    entry = gdal.GetNextDirEntry(d)
+    if entry:
+        gdaltest.post_reason('fail')
+        return 'fail'
+    gdal.CloseDir(d)
+
+    # Depth 1
+    files = [l_entry.name for l_entry in gdal.listdir('/vsimem/vsifile_opendir', 1)]
+    if files != ['subdir', 'subdir/subdir2', 'test']:
```

Comment:
Previous tests also assume a particular order to be returned. That order depends on the fact that /vsimem/ has a std::map (so sorted), and the way GetNextEntry() iterates.  For now, I'll stick with that,but indeed in case of implementation changes, that would have to be revised